### PR TITLE
Test basic AI model

### PR DESCRIPTION
**Pull request by marco-c**
@mbanon what do you think of an approach like this?
Basically, building a ML model where the input features are the top three predictions of fastText and the number of words in dictionaries for these top three predictions.

It might also be worth exploring a more complex model, e.g. increasing the number of "top" predictions to consider (though the more we add, the more expensive the model becomes to run, as it will have to check many dictionaries).

Instead of the number of words, we could also experiment with other features:
- is there at least one word in language X?
- are there **only** words in language X?
- what is the distance to a correct sentence (using https://github.com/MSeal/cython_hunspell#suggestions)? This would also be more robust to typos.

I tested this model on the dataset on https://github.com/pemistahl/lingua-py for Italian and it fared pretty well compared to all other alternatives, especially for shorter sentences. Plain fastspell also fared pretty well (https://github.com/pemistahl/lingua-py/pull/190).

My exploration was done in the context of https://github.com/mozilla/firefox-translations-training/issues/248#issuecomment-1807787199.

### CODE REVIEW

File: src/fastspell/ai.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        else:
+            lang2_dic_tokens = None
+
+        sentences.append({
+            "fastText_lang0": lang0_prob,
+            "fastText_lang1": lang1_prob,
+            "fastText_lang2": lang2_prob,
+            "lang0_dic_tokens": lang0_dic_tokens,
+            "lang1_dic_tokens": lang1_dic_tokens,
+            "lang2_dic_tokens": lang2_dic_tokens,
```
Comment (marco-c) [inline src/fastspell/ai.py:165]: These are the input features for the model.

File: src/fastspell/ai.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        })
+        labels.append(label)
```
Comment (marco-c) [inline src/fastspell/ai.py:167]: The label is 1 if the language to choose is the first among the three, 2 if it's the second, 3 if it's the third.

File: src/fastspell/ai.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+print(len(sentences))
+
+dict_vectorizer = DictVectorizer()
+X = dict_vectorizer.fit_transform(sentences)
+
+xgb_model = xgboost.XGBClassifier(n_jobs=10)
+
+clf = GridSearchCV(
+    xgb_model,
+    {"max_depth": [1, 2, 4, 6], "n_estimators": [25, 50, 100, 200]},
```
Comment (marco-c) [inline src/fastspell/ai.py:182]: We could presumably introduce another hyperparameter, a "threshold" above which we don't need to go look at the dictionary (e.g. if fastText is 99% sure it's Italian, we don't need to check further). This could lower the cost of the approach.

File: src/fastspell/fastspell.py [hunk 4]
```diff
@@ -105,10 +110,36 @@ def search_hunspell_dict(self, lang_code):
                     logging.error("Failed building Hunspell object for " + lang_code)
                     logging.error("Aborting.")
                     exit(1)
-        else:
-            raise RuntimeError(f"It does not exist any valid dictionary directory"
-                               f"for {lang_code} in the paths {self.hunspell_paths}."
+
+        if hunspell_obj is None:
```
Comment (marco-c) [inline src/fastspell/fastspell.py:114]: This change is unrelated, I needed it to load as many dictionaries as possible from my system (as I was testing with Italian, and it is not available in fastspell-dictionaries).

Comment (marco-c) [reply src/fastspell/fastspell.py:114]: I guess this change could be useful in its own right, so users of fastspell can more easily load more dictionaries.

File: src/fastspell/fastspell.py [hunk 5]
```diff
@@ -127,7 +158,7 @@ def load_hunspell_dicts(self):
         self.similar = []
         for sim_entry in self.similar_langs:
             if sim_entry.split('_')[0] == self.lang:
-                self.similar.append(self.similar_langs[sim_entry])
+                self.similar.append(self.similar_langs[sim_entry] + [self.lang])
```
Comment (marco-c) [inline src/fastspell/fastspell.py:161]: This is also unrelated to the AI model of course (this part is not used for it).
I introduced this change to make sure the target language was also considered for spellchecking. Basically, to avoid having to remember to do things like https://github.com/mbanon/fastspell/blob/70d832a2840d9a9f66d873beeaabf643889faa4f/src/fastspell/config/similar.yaml#L4 (where you have to remember to repeat the code of the target language in the similar list: "az: [tr, az]"). I think it is a bit of a footgun for users of the library: as I was testing something like "it: [es, pt, fr, en]", I didn't realize at first without looking at the code that I had to repeate "it".

Interesting thing I didn't expect: in the lingua-py dataset with Italian, when fastText makes a mistake it most often detects Italian as English (https://github.com/pemistahl/lingua-py/blob/534528d99fd0914bdeda92138a58f6073e1067b9/accuracy-reports/fasttext/Italian.txt#L1).

Comment (marco-c) [reply src/fastspell/fastspell.py:161]: Let me know if you want to land this change and I'll open a PR for it.

File: src/fastspell/fastspell.py (hunk 4 above)
Comment (marco-c) [reply src/fastspell/fastspell.py:114]: Let me know if you want to land this change (after cleaning it up of course) and I'll open a PR for it.

File: src/fastspell/fastspell.py (hunk 5 above)
Comment (ZJaume) [reply src/fastspell/fastspell.py:161]: I thought many times of adding this. Please submit a separated PR.

File: src/fastspell/fastspell.py (hunk 4 above)
Comment (ZJaume) [reply src/fastspell/fastspell.py:114]: I'm not sure if I'm not understanding this change correctly, but there's already a whole path search for possible dictionary candidates other than the fastspell-dictionaries if a user wants to use the system's. It's explained in the [documentation](https://github.com/mbanon/fastspell#hunspellyaml) and [here](https://github.com/mbanon/fastspell/blob/main/src/fastspell/util.py#L68) is the code.

File: src/fastspell/config/hunspell.yaml [hunk 6]
```diff
@@ -18,6 +18,7 @@ hunspell_codes:
     da: da_DK
     de: de_DE
     en: en_GB
+    el: el_GR
```
Comment (ZJaume) [inline src/fastspell/config/hunspell.yaml:21]: This will be cleaned up, right? We prefer to add new languages to the default config if the come along with the corresponding fastspell-dictionaries update that adds them.

File: src/fastspell/ai.py (hunk 3 above)
Comment (ZJaume) [reply src/fastspell/ai.py:182]: That's right. I've always been interested if there is a correlation between confidence and precision. But I really never had time to see if the number of false positives or false negatives is very small when confidence is high. Therefore we could add this exception and speed it up.

**Review by ZJaume — COMMENTED** [review]
Hi Marco! I'm glad FastSpell is useful to you and you offer to contribute. This machine learning approach seems pretty interesting and I would like to see some numbers. These are only preliminary comments, as we will need to see more in order to decide if we approve this. But it looks promising!

Regarding the features
> * is there at least one word in language X?
> 
> * are there **only** words in language X?
> 
I like these
> * what is the distance to a correct sentence (using https://github.com/MSeal/cython_hunspell#suggestions)? This would also be more robust to typos.
But this one I'm doubtful if it is expensive.

Also, I would like to suggest a more flexible set of features than this:
https://github.com/mbanon/fastspell/blob/440da84ccfd4498006e5e81c8a804111bd85e42e/src/fastspell/ai.py#L160-L165

Our general experience with the number of dictionaries checking for errors, is that 4 is starting to be expensive and 5 or more is generally prohibitive.
Of course, this always depends on if the language we are trying to improve is low-resource or not. For example, Icelandic is low-resource, we are interested to preserve as much as data as possible and the amount of data we usually process for Icelandic is low. So using 5 dictionaries for Icelandic is not a big deal. But it may be for Spanish or Italian.

Currently the number of similars is different in every set of languages and we'd like to keep it that way for the reasons I explained before. So I think the features should allow this. Then, I also found many times that the correct language is sometimes like 7th or 8th in the top predictions, so it could be necessary to have a number of features `*fasttext_prob` different (and bigger) than the `*dic_tokens`.

After thinking about this, the idea that comes to my mind it would be to use a feature array that each position represents one of the all possible labels fasttext has (in the model we use, would be 176). So the position 0 represents Afrikaans probability, the position 1 AbKhasian and so on. Then, if we extract, for example, the top10 languages, we set for those languages in the array, the corresponding probability value. For the rest just 0. For the number of correct tokens in each language, it would be more or less the same. Another 176 features and set the corresponding positions to the values that we obtained after hunspell checking.

It this could be easily achived with https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html

The idea behind this is that the model could be able to establish the relationship between each feature and label, so in the end you could give to it any arbitrary number of probabilities and spellcheck corrections.
Of course, this could be an overkill and maybe too expensive to run on each segment to classify. But I think the features are still cheap to compute this way. Maybe the classifier with 260 features is not.


A more simpler approach could be simply having this?
```
            "fastText_lang0": lang0_prob,
            "fastText_lang1": lang1_prob,
            ...
            "fastText_lang9": lang1_prob,
            "lang0_dic_tokens": lang0_dic_tokens,
            "lang1_dic_tokens": lang1_dic_tokens,
            "lang2_dic_tokens": lang2_dic_tokens,
            "lang2_dic_tokens": lang2_dic_tokens,
 ```
and always giving the top10 and setting to -1 the `dic_tokens` when you haven't checked that language? But in this case the classifier desn't know to which language of the spellings corresponds which probability.

EDIT: I just saw that you approach is to predict which of those 3 feats is the correct language, not predicting the language label. So maybe my idea is not useful anymore. But I still think more flexibility could improve the model.

File: src/fastspell/ai.py (hunk 2 above)
Comment (ZJaume) [reply src/fastspell/ai.py:167]: Oh sorry, I did not see this. Maybe this changes a little bit my proposal.

**Comment by ZJaume** [comment]
I also have the doubt of what happens when the language we are targetting is not supported in FastText (therefore will never appear in the topk) or in other extreme cases. Leike Uzbek #14 that indicates that probably fasttext never had any cyrillic training for Uzbek so when it comes Uzbek in cyrillic it would also not appear in the topk.

File: src/fastspell/fastspell.py (hunk 4 above)
Comment (marco-c) [reply src/fastspell/fastspell.py:114]: This was to allow loading a dictionary more easily, e.g. if "it" is the language and there's "it-IT.dic" in the system, it will assume that's the one to load (even though there might be others like "it_CH.dic").

File: src/fastspell/fastspell.py (hunk 5 above)
Comment (marco-c) [reply src/fastspell/fastspell.py:161]: I opened #20.

File: src/fastspell/config/hunspell.yaml (hunk 6 above)
Comment (marco-c) [reply src/fastspell/config/hunspell.yaml:21]: Yeah this was just for testing.

**Comment by marco-c** [comment]
> EDIT: I just saw that you approach is to predict which of those 3 feats is the correct language, not predicting the language label. So maybe my idea is not useful anymore. But I still think more flexibility could improve the model.

Your proposal is very similar to what I wanted to try too:
> It might also be worth exploring a more complex model, e.g. increasing the number of "top" predictions to consider (though the more we add, the more expensive the model becomes to run, as it will have to check many dictionaries).
with the addition of the confidence threshold.

I think the approach is promising, unfortunately I won't have a lot of time to devote to the exploration soon. It might be an interesting research project for a student or an intern :)

**Comment by marco-c** [comment]
Yet another option, which would be more costly to train but likely provide better results, is training a fastText model with the addition of dictionary features, so everything is learned together instead of building a model on top of another one.

I guess we could use https://github.com/laurieburchell/open-lid-dataset to train and validate the model (either the simple one on top of fastText or the more complex one which would fully replace fastText).
